storage/graphdriver/rawblock/driver.go

```diff
@@ -0,0 +1,245 @@
+package rawblock
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/docker/docker/daemon/graphdriver"
+	"github.com/docker/docker/pkg/idtools"
+	"github.com/golang/glog"
+	"github.com/opencontainers/runc/libcontainer/label"
+	//"github.com/docker/docker/pkg/mount"
+)
+
+func init() {
+	graphdriver.Register("rawblock", Init)
+}
+
+// Driver holds information about the driver, home directory of the driver.
+// Driver implements graphdriver.ProtoDriver. It uses only basic vfs operations.
+// In order to support layering, the block is created via reflink with the the parent layer
+// Driver must be wrapped in NaiveDiffDriver to be used as a graphdriver.Driver
+type Driver struct {
+	home      string
+	backingFs string // host filesystem of the storage
+	cow       bool
+	blockFs   string // filesystem inside the block
+	blockSize int    // block size in GB
+	uidMaps   []idtools.IDMap
+	gidMaps   []idtools.IDMap
+
+	sync.Mutex // Protects concurrent modification to active
+	active     map[string]int
+}
+
+// Init returns a new Raw Block driver.
+// This sets the home directory for the driver and returns NaiveDiffDriver.
+func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (graphdriver.Driver, error) {
+	backingFs := "<unknown>"
+	blockFs := "xfs" // TODO: make it configurable
+	cow := true
+	supported := "supported"
+
+	fsMagic, err := graphdriver.GetFSMagic(home)
+	if err != nil {
+		return nil, err
+	}
+	if fsName, ok := graphdriver.FsNames[fsMagic]; ok {
+		backingFs = fsName
+	}
+
+	// check if they are running over btrfs or xfs
+	switch fsMagic {
+	case graphdriver.FsMagicBtrfs: // support
+	case graphdriver.FsMagicXfs: // TODO: check support(201609) and use ioctl(FICLONE) instead of cp --reflink
```

Comment:
it supports all host filesystem. only btrfs supports cow. xfs supports cow soon.